### ISSUE TITLE
Search File Aliases

### ISSUE THREAD

**Issue by Khoding** [AlexBieg/obsidian-better-command-palette#32]
I tried to do it myself but Idk enough about both Typescript and Obsidian's API, so here I am instead 😅

I often use aliases while using the file searching utility in Obsidian, and I feel like this could entirely replace the whole implementations of both file search and command palette if it gave the ability to search file aliases.

As a default prefix I was thinking @ would be great, but I think it's already used somewhere? I guess you'll find something anyway.

I also thought maybe it'd be even better if we use File search prefix + @ or something, it focuses on aliases, else it works like the normal file search.

And maybe I'm thinking too much but I believe it'd be great if we could toggle this part in the settings

**Comment by iamrecursion** [comment]
I'd just like to add my +1 for being able to search frontmatter aliases as filenames. The native open dialog does so, and it's pretty much the only reason I still use that sometimes!

**Comment by AlexBieg** [comment]
Alrighty, the file alias search is out. I'm adding new issues for searching the tag and the unstructured frontmatter content, but for now, this should hopefully unblock you all. One small note, I just grouped the aliases with the normal file name (as opposed to how obsidian does it of having each alias be a separate entry). This was mostly due to  my current architecture making this approach easier and I don't anticipate any real issues, but let me know and we can discuss alternatives.

**Comment by iamrecursion** [comment]
Perhaps I'm being a touch silly, but I can't get this to work as expected on mobile. 

Just to clarify, I'm opening the command palette, hitting `/` and then searching for an alias. It seems like they don't always appear, and when they do the ranking is very odd with them appearing low on the list even if the alias is a much better match than things ranked higher (e.g. file with raw name "Foo Bar" appears above file "baz" with exact alias "Foo" when searching for "Foo").

**Comment by AlexBieg** [comment]
@iamrecursion Interesting, are you seeing different behavior on mobile compared to desktop? Theoretically, they should work the same, but I might be missing some nuances of the mobile environments. If it is working differently could you give me your mobile OS and version?

Regarding search order, that's a bit more complicated to fix. The way the fuzzy search works currently is on a single string per file. For example, if I have a file `baz.md` and it has an alias `Foo` the string that is fuzzy searched is `baz.md:Foo`. The reason you are seeing that ordering is that for at least this fuzzy search implementation it's ranking `Foo Bar.md` higher than `baz.md:Foo` probably because the first one has a match earlier in the string. Obviously, we could reverse the order of names and aliases, but that leads to the same problem, but in reverse. This is probably the reason Obsidian has their aliases split into separate suggestions. Honestly that is probably the move here too. I was being a bit lazy by just appending them on. I'll re-open this card since this really isn't working as should be. The future version of this will have those aliases as separate suggestions that should be ordered correctly.

**Comment by iamrecursion** [comment]
I think you're probably right about the separate suggestions.

I only mentioned mobile because it was the one platform I'd tested it on! I'm now at my computer and I'm observing the same behaviour, so there don't seem to be any differences between platforms. 

Your explanation makes perfect sense for some of the more "out there" results that I'm seeing, where things like matching the string "foo" across separate words in the title still ranks higher than the alias!

**Comment by AlexBieg** [comment]
Just pushed out the patch for that weird search behavior. Now aliases are their own suggestion, just like the standard switcher does. The search ordering should be much better now 😄

**Comment by iamrecursion** [comment]
Sorry to be the bearer of bad news, but there seems to be a bug here! When there are multiple aliases that are the same, only one of them shows up. If I have a note "James X" and a note "James Y", both aliased to "James", only one shows up. 

Along similar lines, the way that the standard open dialog shows the file being aliased below the alias is important for disambiguating multiple files with the same alias, and is very useful in and of itself for context. 

On a positive note, it has immediately improved the search behaviour for aliases!

**Comment by AlexBieg** [comment]
> Sorry to be the bearer of bad news, but there seems to be a bug here! When there are multiple aliases that are the same, only one of them shows up. If I have a note "James X" and a note "James Y", both aliased to "James", only one shows up.

@iamrecursion Boy howdy this was a hard one to track down. Turns out the place I was accessing all of the files was a part of the cache that (for some reason unknown to me) is updated asynchronously from the rest. So on a reload or whenever the cache is rebuilt that section would take a long time to populate causing the issue. This was fixed by just pulling the file paths from a different part of the cache.

> Along similar lines, the way that the standard open dialog shows the file being aliased below the alias is important for disambiguating multiple files with the same alias, and is very useful in and of itself for context.

Makes a ton of sense and honestly as someone who doesn't really use aliases I did not realize you could even have duplicates. Easy fix. I changed the UI slightly from how the native Palette does it because they put it is where I store the tags. Hopefully, this helps with the issues you were running into.

Those fixes were released in 0.11.2

**Comment by iamrecursion** [comment]
Oh goodness I really do seem to be the bearer of bad news today! 0.11.3 does fix the issue where the list would fail to populate properly from 0.11.2, but there are still a couple of issues that somewhat break things:

1. A lot of aliases are missing from the searcher. Even ones I've had for ages just plain don't show up under the searcher, but others do. Very strange.
2. I still can't find any way to make two aliases of the same name show up at once.

Perhaps I'm running into some issues due to my mammoth vault again?

**Comment by AlexBieg** [comment]
@iamrecursion That is very strange. I'll keep digging into this. Sorry that this keeps being an issue. I think this ticket has shown the need for this plugin to have tests. At this point, there are enough features/edge cases that I'm not effectively testing manually. Once I get this ticket settled that will be my next priority so that I can push out new features with a bit more confidence.

**Comment by AlexBieg** [comment]
@iamrecursion I think this is likely related to the size of your vault, but I'm not positive. I'm having a hard time replicating these issues. My test vault has 1000 files all with the same alias and one unique to each file and while the first indexing pass takes a second to populate everything, subsequent reloads seem speedy. I have three questions/requests:

1. I'm curious if you would be able to create a quick test vault with fewer files in it and see if it's still happening?
    - If it does happen in that vault please send me the repro steps and I'll recreate it in my test env
2. Is this on mobile? I'm wondering if maybe the cache population takes longer on slower devices
3. If you see that a certain alias is not listed, can you try opening that file and see if it is then listed? From my understanding opening a file pushes it to the front of the cache queue so it would be another indication that your cache population is just taking a long time

Without being able to reproduce this I don't have a lot of paths forward, unfortunately. That said I'm happy to keep working on it when I have more info

**Comment by iamrecursion** [comment]
> 1. I'm curious if you would be able to create a quick test vault with fewer files in it and see if it's still happening?

I've managed to create one with only two files and it still appears to happen. Please find it [here](https://github.com/AlexBieg/obsidian-better-command-palette/files/8103063/Test.zip).

1. Load the vault
2. Hit `⌘ + ⇧ + P` to open BCP.
3. Type `/James`.
4. Neither of the aliases show up, just the files.

This happens no matter how many times I reload the vault for me.

> 2. Is this on mobile? I'm wondering if maybe the cache population takes longer on slower devices

I hadn't actually checked on mobile before now, but it does reproduce.

> 3. If you see that a certain alias is not listed, can you try opening that file and see if it is then listed? From my understanding opening a file pushes it to the front of the cache queue so it would be another indication that your cache population is just taking a long time

Opening the file makes no difference.

**Comment by iamrecursion** [comment]
> Sorry that this keeps being an issue.

Honestly, it's okay! I find that I tend to break plugins in unexpected ways; 'tis my curse! I'm always up to help as best I can! I only lament that I don't have the time to actually dig in from a code perspective at the moment. 

That said, if there are any queries that I could run in the console that would help debug things, please do let me know!

**Comment by AlexBieg** [comment]
@iamrecursion @Khoding Okay I think 🤞 this should be working as expected now. It turns out that both `alias` and `aliases` are acceptable frontmatter keys. I was only checking the latter. I have learned about a utility provided by the obsidian API parses the aliases for me, so that is all handled nicely now. As a note (and this is true for the standard quick switcher as well) if you specify both `alias` and `aliases` in the frontmatter block it will only respect the first one in the block. I will once again be optimistically closing this 😄 

I'd also like to apologize for the delay. I spent quite a bit of time getting some automated UI testing in place. Hopefully, it should reduce a lot of the regression bugs as well as my testing time.
